### PR TITLE
assignment 5 logic in functional programming format

### PR DESCRIPTION
**Pull request by AritraChatterjee-SDE-SWE**
implemented the assignment five logic.

next sub-task:
1. convert to oops based code.
2. add error handling

future task:
1. Add rspec for unit testing

### CODE REVIEW

File: Assignment_05.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+1. Define required class & methods for it
+2. split down the main method into multiple methods meaningfully
+3. validate input & handle exception for invalid input
+4. write unit test cases for this class functionalities using rspec
+=end
+
+require 'date'
+
+# This method generates and returns the output stuctute
+def gen_output_structure()
```
Comment (qisantanu) [inline Assignment_05.rb]: Please create a Class which will maintain the validation and required processing part

**Review by qisantanu — CHANGES_REQUESTED** [review]

File: Assignment_05.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+=end
+
+require 'date'
+
+class TaskHandler
+  def initialize(input)
+    @input = input
+  end
+
+  def validate
```
Comment (qisantanu) [inline Assignment_05.rb]: should it be a private method?

**Review by qisantanu — CHANGES_REQUESTED** [review]

File: Assignment_05.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  
+1. Define required class & methods for it
+2. split down the main method into multiple methods meaningfully
+3. validate input & handle exception for invalid input
+4. write unit test cases for this class functionalities using rspec
+=end
+
+require 'date'
+
+class TaskHandler
```
Comment (qisantanu) [inline Assignment_05.rb]: name it as `Task`

File: Assignment_05.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    { id: 6, name: "task6", time_logs: [[DateTime.parse("2023-01-01 17:30:00"), DateTime.parse("2023-01-01 18:30:00")]] },
+  ]
+
+
+expected output format:
+
+  {
+    categories: ["task1", "task2", "task3", "task4", "task5", "task6"],
+    total_time: [<total_spent_time_in_minutes_for_task1>,  .. <total_spent_time_in_minutes_for_task6>],
+    total_time: [<(total_spent_time_in_minutes_for_task_1 / (24 * 60)) * 100>, .. <(total_spent_time_in_minutes_for_task_6 / (24 * 60)) * 100>]
```
Comment (qisantanu) [inline Assignment_05.rb:31]: the key 
should be different
occupation_percecntage

File: Assignment_05.rb (hunk 3 above)
Comment (qisantanu) [reply Assignment_05.rb]: We can rearrange the code like this:

  ```
class TaskHandler
  - will accept the user inputs
  - will create Task Object
  - def process/ process_input
  - #validation
  -   - validate the user input


class Task
  - id
  - name
  - time_logs
  
  - def total_time
  - def occupation_percecntage_day_wise
  
 # validation pvt methods
  - validate_name
  - validate_time_logs   
  
  ```

**Review by qisantanu — CHANGES_REQUESTED** [review]

File: Assignment_05.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    time_logs.sum { |time_log| get_time_duration_in_minute(time_log[0], time_log[1]) }.round()
+  end
+
+  # This method takes total time used for a task and returns the in percentage form for a day
+  def occupation_percecntage_day_wise(total_time_in_minute)
+    (total_time_in_minute.to_f / (24 * 60) * 100).round(2)
+  end
+
+  # This method validate the the time logs
+  def validate_time_logs
```
Comment (qisantanu) [inline Assignment_05.rb:76]: should be under private access

Comment (AritraChatterjee-SDE-SWE) [reply Assignment_05.rb:76]: I am using the validate_time_logs outside the Task class, that is why it is public. I am working on to make it private.

File: Assignment_05.rb [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    time_logs.sum { |time_log| get_time_duration_in_minute(time_log[0], time_log[1]) }.round()
+  end
+
+  # This method takes total time used for a task and returns the in percentage form for a day
+  def occupation_percecntage_day_wise(total_time_in_minute)
+    (total_time_in_minute.to_f / (24 * 60) * 100).round(2)
+  end
+
+  # This methods creates a obj of Task class and returs it after validation else nil
+  def add(id="deafult", name="deafult", time_logs=[])
```
Comment (qisantanu) [inline Assignment_05.rb]: the method should not be here.

File: Assignment_05.rb [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    some_task = Task.new.add(a_task[:id], a_task[:name], a_task[:time_logs])
+    if some_task
+      @tasks << some_task
+    else
+      puts "Task with id: #{a_task[:id]} not added"
+    end
+  end
+
+  def add_many_task(tasks)
+    tasks.each do |a_task|
```
Comment (qisantanu) [inline Assignment_05.rb]:
```
tasks.each do |a_task|
    task_object = Task.new(a_task[:id], a_task[:name], a_task[:time_logs])
    
    if task_object.valid?
      @tasks << some_task
    else
      puts "Task with id: #{a_task[:id]} not added"
    end
  end
```

File: Assignment_05.rb [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  # This methods creates a obj of Task class and returs it after validation else nil
+  def add(id="deafult", name="deafult", time_logs=[])
+    a_task = Task.new(id, name, time_logs)
+
+    if validate_time_logs(id, name, time_logs)
+      a_task
+    else
+      nil
+    end
+  end
```
Comment (qisantanu) [inline Assignment_05.rb:65]:
```
def valid?
    validate_time_logs
end

```

File: Assignment_05.rb [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+  private
+  
+  # This methods takes an time log array as argument [start_time, end_time] and retures the duration in minute
+  def get_time_duration_in_minute(start_time, end_time)
+    duration = DateTime.parse(end_time) - DateTime.parse(start_time)
+    duration.to_f * 24 * 60
+  end
+
+  # This method validate the the time logs
+  def validate_time_logs(id, name, time_logs)
```
Comment (qisantanu) [inline Assignment_05.rb]: no need to pass this.
self.id/ self.name

File: Assignment_05.rb [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+end
+
+class TaskHandler
+  attr_reader :tasks
+
+  def initialize
+    @tasks = []
+  end
+
+  def add_one_task(a_task)
```
Comment (qisantanu) [inline Assignment_05.rb]: rename it to `task_attributes`

File: Assignment_05.rb (hunk 7 above)
Comment (qisantanu) [reply Assignment_05.rb]: reame it to `task_attr_array`

**Review by qisantanu — CHANGES_REQUESTED** [review]

File: Assignment_05.rb [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+  def initialize(id="deafult", name="deafult", time_logs=[])
+    @id = id
+    @name = name
+    @time_logs = time_logs
+  end
+
+  # This method returns the total time used for the task
+  def total_time_in_minute
+    time_logs.sum { |time_log| get_time_duration_in_minute(time_log[0], time_log[1]) }.round()
```
Comment (qisantanu) [inline Assignment_05.rb]: remove the parenthesis from the round method

File: Assignment_05.rb [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+class TaskHandler
+  attr_reader :tasks
+
+  def initialize
+    @tasks = []
+  end
+
+  def add_one_task(task_attributes)
+    task = Task.new(task_attributes[:id], task_attributes[:name], task_attributes[:time_logs])
+    if task.valid?
```
Comment (qisantanu) [inline Assignment_05.rb:99]: please give line break when you are starting a block, and no indentation is there.

**Review by qisantanu — APPROVED** [review]
:rocket:

**Pull request merged**